### PR TITLE
Rs/wx/rule5 45

### PR DESCRIPTION
**Pull request by weilixu**
Update the 5-45

One issue is the schedules are at the ruleset_model_instance level.

So I have to create a nested class of RMD to extract the schedules.

I prefer this implementation because the RMD is still not approved by the full committee, and it is subject to change in the future.

### CODE REVIEW

**Comment by weilixu** [comment]
Also, ignore the JSON test changes - once Juan's test merges, those will automatically be gone.

File: rct229/rules/section5/section5rule45.py [hunk 1]
```diff
@@ -16,57 +17,81 @@ class Section5Rule45(RuleDefinitionListIndexedBase):
...
+                    next((getattr_(sched, "schedule", "hourly_values") for sched in schedules_b if sched["id"] == infiltration_b["multiplier_schedule"]), []),
+                    next((getattr_(sched, "schedule", "hourly_values") for sched in schedules_p if sched["id"] == infiltration_p["multiplier_schedule"]), []),
+                    mask_schedule,
+                    1.0,
+                    data["is_leap_year"],
+                )
+
+            def rule_check(self, context, calc_vals=None, data=None):
+                total_hours_compared = calc_vals["total_hours_compared"]
+                total_hours_match = calc_vals["total_hours_matched"]
```
Comment (charliepnnl) [inline rct229/rules/section5/section5rule45.py:88]: Should this be `total_hours_matched = ...`?

File: rct229/rules/section5/section5rule45.py [hunk 2]
```diff
@@ -16,57 +17,81 @@ class Section5Rule45(RuleDefinitionListIndexedBase):
...
+                return (
+                    total_hours_compared > 0.0 and total_hours_compared == total_hours_match
+                )
 
-        def get_fail_msg(self, context, calc_vals=None, data=None):
-            elfh_difference = calc_vals["elfh_difference"]
-            return f"Baseline and proposed infiltration schedules are not the same. The difference of baseline EFLH to proposed EFLH is equal to {elfh_difference}."
+            def get_fail_msg(self, context, calc_vals=None, data=None):
+                elfh_difference = calc_vals["eflh_difference"]
+                return f"Baseline and proposed infiltration schedules are not the same. " \
```
Comment (charliepnnl) [inline rct229/rules/section5/section5rule45.py:96]: Please update the RDS to reflect this message wording change.
BTW: I think it should be "... The difference between baseline EFLH and proposed EFLH is ..."

**Review by charliepnnl — APPROVED** [review]
Approved with mostly stylistic comments.

File: rct229/rules/section5/section5rule45.py [hunk 3]
```diff
@@ -16,57 +17,81 @@ class Section5Rule45(RuleDefinitionListIndexedBase):
...
-                LEAP_YEAR_HOURS if data["is_leap_year"] else REGULAR_YEAR_HOURS
-            )
+        def list_filter(self, context_item, data=None):
+            # baseline should match to proposed
+            zone_b = context_item.baseline
+            return "infiltration" in zone_b
 
-            infiltration_p = zone_p["infiltration"]
-            infiltration_b = zone_b["infiltration"]
+        class ZoneRule(RuleDefinitionBase):
```
[the remark this answers is not in the document]
Comment (weilixu) [reply rct229/rules/section5/section5rule45.py:54]: The RDS group here is to extract `schedules` in `create_data` function.

I don't see the need for the BuildingRule other than output... From the post-processing point of view, do you think we should enforce building rule level for every rule?

File: rct229/rules/section5/section5rule45.py [hunk 4]
```diff
@@ -16,57 +17,81 @@ class Section5Rule45(RuleDefinitionListIndexedBase):
...
+                    "schedules_b": rmd_b["schedules"],
+                    "schedules_p": rmd_p["schedules"]}
 
-            mask_schedule = [1.0] * (
-                LEAP_YEAR_HOURS if data["is_leap_year"] else REGULAR_YEAR_HOURS
-            )
+        def list_filter(self, context_item, data=None):
+            # baseline should match to proposed
+            zone_b = context_item.baseline
+            return "infiltration" in zone_b
```
[the remark this answers is not in the document]
Comment (weilixu) [reply rct229/rules/section5/section5rule45.py:52]: Correct. Putting `infiltration` in the required_fields shall work. I will update it.

File: rct229/rules/section5/section5rule45.py (hunk 2 above)
Comment (weilixu) [reply rct229/rules/section5/section5rule45.py:96]: You are correct. Updated

File: rct229/rules/section5/section5rule45.py [hunk 5]
```diff
@@ -16,57 +17,81 @@ class Section5Rule45(RuleDefinitionListIndexedBase):
...
+                )
+
+                infiltration_b = zone_b["infiltration"]
+                infiltration_p = zone_p["infiltration"]
+
+                return compare_schedules(
+                    # search for the schedule used in infiltration data group.
+                    # Set to empty list if non-matched,
+                    # raise exception if not hourly schedule or the hourly_values key is missing
+                    next((getattr_(sched, "schedule", "hourly_values") for sched in schedules_b if sched["id"] == infiltration_b["multiplier_schedule"]), []),
```
[the remark this answers is not in the document]
Comment (weilixu) [reply rct229/rules/section5/section5rule45.py:79]: Great idea! I included the changes in the latest push

File: rct229/rules/section5/section5rule45.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (weilixu) [reply rct229/rules/section5/section5rule45.py:52]: @charliepnnl Addressed all the comment but this one. Let me know if it makes sense.

**Pull request merged**